### DISCUSSION TITLE
Trying to understand Pyright behavior resolving partial stub - Marshmallow

### DISCUSSION THREAD

**Discussion by dephiros** [microsoft/pyright#6063] [General]
**Reproducible example**: https://github.com/dephiros/pyright-marshmallow-all
https://vimeo.com/manage/videos/870388634/cce97ff508
**What I am trying to do**: modify the type for a few method inside `marshmallow/schema.py` module by adding a stub at typings/marshmallow/schema.pyi(complete module stub)

**Expected result**: Pyright are able to recognize the stubs and merge it with other modules inside marshmallow

**What I see instead:** 
```
/Users/annguyen/projects/pyright-marshmallow-all/src/__init__.py
  /Users/annguyen/projects/pyright-marshmallow-all/src/__init__.py:9:30 - error: "Schema" is not a known member of module "marshmallow" (reportGeneralTypeIssues)
  /Users/annguyen/projects/pyright-marshmallow-all/src/__init__.py:6:23 - error: "ValidationError" is not a known member of module "marshmallow" (reportGeneralTypeIssues)
  /Users/annguyen/projects/pyright-marshmallow-all/src/__init__.py:26:31 - error: "ValidationError" is not a known member of module "marshmallow" (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations 
```
This error requires adding an `typings/marshmallow/__init__.pyi`

**Source code for reference:**
- [Attempt 1: try only `typings/marshmallow/schema/__init__.pyi`](https://github.com/dephiros/pyright-marshmallow-all/commit/77d9699432791aa46ce719e43dde6f3d2da292a2)
- [Attempt 2: try only `typings/marshmallow/schema.pyi`](https://github.com/dephiros/pyright-marshmallow-all/commit/8eeae4564133996f28847ea5b0cae2432a63aa87)
- [Attempt 3: try both `typings/marshmallow/schema.pyi` and `typings/marshmallow/__init__.pyi`](https://github.com/dephiros/pyright-marshmallow-all/commit/e028664e0b1173bae754c2decabee788d0bcc576)

**Questions**: Would someone be able to explain why `typings/marshmallow/__init__.pyi` and what would be the recommend approach?

**Comment by erictraut** [comment]
Overriding specific submodules within the library with stubs is not going to work well, as you've discovered. I recommend generating stubs for the entire library and then editing or augmenting stub files for the specific submodules that you want to change.

**Reply by sam-goodwin** [reply]
This is highly inconvenient and intrusive. Just to patch one function I have to copy the entire package into my repo? Why can't declaration merging be applied like is done in languages like typescript?

**Comment by dephiros** [comment]
@erictraut , thank you for you response 🙏
A few more question so I can understand better:
- This section in [PEP-0561 mentions](https://peps.python.org/pep-0561/#partial-stub-packages)

> Type checkers should merge the stub package and runtime package or Typeshed directories. This can be thought of as the functional equivalent of copying the stub package into the same directory as the corresponding runtime package

[Also see in a Github thread that partial stub should work at the module level](https://github.com/python/mypy/issues/5880)

Which lead me to believe that there is a fallback behavior. From my experiment above, it does work to a certain extend but not what I expected.

1. Why does the merging behavior not work?
2. Does Pyright aligns PEP-0561 in this case?
3. Should Pyright support partial stubbing of package to address this use case?

**Reply by sam-goodwin** [reply]
I would like an answer to this as well. I don't think duplicating the entire module is a good solution. pyright provides a CLI to do it but still, declaration merging should be possible.

**Reply by erictraut** [reply]
Yes, I believe that pyright conforms to PEP 561. If you see a behavior that looks like it doesn't conform to PEP 561, please let us know.

There is no provision in the typing spec for "merging" type information contained within a type stub and a module's implementation — or multiple stubs — to produce some transformed view of type information for a module.

Pylance attempts to do something akin to this to merge class/function docstrings (which are often omitted from stub files) with the type information in stubs. This process is very error prone and relies on some sketchy heuristics. It works well enough for docstring merging because the consequences of failure are small. For type checking, the process would need to be fully deterministic with clear rules that all type checkers can adhere to.

Type stubs were not designed with this use case in mind, so it's not surprising that this wouldn't work well.

If you're interested in exploring a solution to this problem, you're welcome to start a discussion in the [Python typing forum](https://discuss.python.org/c/typing/32). I think you may be underestimating the complexity involved in defining rules for merging type information at the module level (or at finer granularity like methods within a class). There are many tricky design decisions involved. If you think this is something you'd like to pursue, it would most likely require a PEP.

**Reply by sam-goodwin** [reply]
Hi @erictraut, I appreciate that there is likely a ton of complexity that is opaque to me and am not claiming it's easy. More looking at it from the point of view of a user. Having to duplicate the entire surface area of a package is not convenient. It's a lot of redundant code that needs to be checked in and maintained as the library changes as opposed to a layered approach like is seen in TypeScript's module augmentation. 

Unfortunate that it would require a PEP to do such a thing. I'll consider starting a thread over there. Thanks for the information.
